client/src/includes/tabs.js

```diff
@@ -0,0 +1,339 @@
+      this.tabButtons.forEach((button) => {
+        button.setAttribute(
+          'aria-controls',
+          button.getAttribute('href').replace('#', ''),
+        );
+      });
+
+      // Check for active items set by the template
+      const tabActive = [...this.tabButtons].find(
+        (button) => button.getAttribute('aria-selected') === 'true',
+      );
+
+      if (window.location.hash && !this.disableURL) {
+        this.selectTabByURLHash();
+      } else if (tabActive) {
+        // If a tab isn't hidden for some reason hide it
+        this.tabPanels.forEach((tab) => {
+          // eslint-disable-next-line no-param-reassign
+          tab.hidden = true;
+        });
+        // Show aria-selected tab
+        this.selectTab(tabActive);
+      } else {
+        this.selectFirstTab();
+      }
+    }
+  }
+
+  /**
+   * @param {string}newTabId
+   */
+  unSelectActiveTab(newTabId) {
+    // IF new tab ID is the current then don't transition out
+    if (newTabId === this.state.activeTabID || !this.state.activeTabID) {
+      return;
+    }
+
+    // Tab Content to deactivate
+    const tabContent = this.tabContainer.querySelector(
+      `#${this.state.activeTabID}`,
+    );
+
+    if (!tabContent) {
+      return;
+    }
+
+    if (this.animate) {
+      this.animateOut(tabContent);
+    } else {
+      tabContent.hidden = true;
+    }
+
+    const tab = this.tabContainer.querySelector(
+      `a[href='#${this.state.activeTabID}']`,
+    );
+
+    tab.setAttribute('aria-selected', 'false');
+    tab.setAttribute('tabindex', '-1');
+  }
+
+  /**
+   * @param {HTMLElement}tab
+   */
+  selectTab(tab) {
+    if (tab) {
+      const tabContentId = tab.getAttribute('aria-controls');
+
+      // Unselect currently active tab
+      if (tabContentId) {
+        this.unSelectActiveTab(tabContentId);
+      }
+
+      this.state.activeTabID = tabContentId;
+
+      tab.setAttribute('aria-selected', true);
+      tab.removeAttribute('tabindex');
+      const tabContent = this.tabContainer.querySelector(`#${tabContentId}`);
+
+      if (!tabContent) {
+        return;
+      }
+
+      if (this.animate) {
+        this.animateIn(tabContent);
+      } else {
+        tabContent.hidden = false;
+      }
+
+      if (this.state.initialPageLoad) {
+        // On first load set the scroll to top to avoid scrolling to active section and header covering up tabs
+        setTimeout(() => {
+          window.scrollTo(0, 0);
+        }, this.state.transition * 2);
+      }
+
+      // Dispatch tab selected event for the rest of the admin to hook into if needed
+      // Trigger tab specific switch event
+      this.tabList.dispatchEvent(
+        new CustomEvent('switch', { detail: { tab: tab.dataset.tab } }),
+      );
+      // Dispatch tab-changed event on the document
+      document.dispatchEvent(new CustomEvent('tab-changed'));
```

Comment:
Can we prefix our custom events? Also I recommend we don't have two events here. The tabList one will bubble to the document. 

Unless we want a cancellable event that fires before switching and one that fires after the switch/change has happened. Either way these should dispatch on the tab container or tabList for both. 

Similar to the title change - using `wagtail:`

E.g.

`wagtail:tab-switch` and `wagtail:tab-changed`

See https://docs.wagtail.org/en/stable/advanced_topics/documents/title_generation_on_upload.html  for an example of other custom events. 

We will also need to sure the release notes include the change from bootstrap and hence their events to new ones. 

These events are not documented but it's good to give people a heads up.

Reply:
I think it is useful to have an event trigger before we are about to switch tabs that can be cancelled. https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault

Then an event that triggers after the tab has changed. 

Both should dispatch from the same element. Maybe the tab button or the container. 

This gives us and others a way to hook into 'i want to block this or change something before the tab is shown ' vs 'i want to do something after the tab has been shown / inc. any transitions'

Reply:
Note. If we want to tackle event naming separately - fully agree it may be out of scope for this PR. 

Plus, we may find this becomes cleaner to manage with a future lightweight framework (if that eventuates)

Reply:
Ok